Add ESLint rule for Metabase strings

WiNloSt opened the pull request:
This is a follow up from #34920, similar to the ESLint rule we introduce for the epic #36966

### Description

This ESLint rule ensures we won't forget to treat user-facing `Metabase` string properly. When users are whitelabeling, any user facing strings that are expected to be seen by non-admins should be replaced with the `application-name` setting instead.

### How to verify

When you reference `Metabase` string. This ESLint rule should start complaining. I exclude some files from this ESLint rules and hope that makes sense.


### Checklist

- [x] Tests have been added/updated to cover changes in this PR

WiNloSt committed:
Fix case where Metabase links are render inside template strings

WiNloSt committed:
Add `no-literal-metabase-strings` ESLint rule

WiNloSt committed:
Fix all Metabase string errors

WiNloSt commented on .eslintrc:
```diff
@@ -161,13 +161,23 @@
     },
     {
-      "files": ["*.unit.spec.*", "frontend/src/metabase/admin/**/*"],
+      "files": [
```
I added more stuff here that should not be linted.

WiNloSt commented on frontend/src/metabase/setup/components/PreferencesStep/PreferencesStep.tsx:
```diff
@@ -73,7 +73,6 @@ export const PreferencesStep = (): JSX.Element => {
       <StepDescription>
         {t`In order to help us improve Metabase, we'd like to collect certain data about product usage.`}{" "}
         <ExternalLink
-          // eslint-disable-next-line no-unconditional-metabase-links-render -- Metabase setup
```
This is now excluded in the `.eslintrc`

WiNloSt commented on frontend/lint/tests/no-unconditional-metabase-links-render.unit.spec.js:
```diff
@@ -131,6 +139,15 @@ const docsLink = useSelector(state =>
+      /Metabase links must be rendered conditionally\.(.|\n)*Please import `getShowMetabaseLinks`(.|\n)*Or add comment to indicate the reason why this rule needs to be disabled/,
+  },
+  {
+    name: 'Detect "metabase.com/docs" in template strings',
```
add missing tests.

WiNloSt commented on frontend/lint/eslint-rules/no-unconditional-metabase-links-render.js:
```diff
@@ -172,7 +172,10 @@ module.exports = {
-          if (LITERAL_METABASE_URL_REGEX.exec(quasi.value.raw)) {
+          if (
+            LITERAL_METABASE_URL_REGEX.exec(quasi.value.raw) &&
+            !isGetShowMetabaseLinksSelectorImported
```
Fixes this rule not working in template literals. I also add relevant tests.

WiNloSt commented on frontend/lint/eslint-rules/no-literal-metabase-strings.js:
This rule is very similar to `frontend/lint/eslint-rules/no-unconditional-metabase-links-render.js`, but it's simpler.

We detect if we have `"Metabase"` in any strings, and show that we could fix the error by either importing `getApplicationName` or disabling the rule with a comment. Because there would be some places where we don't want to use `getApplicationName` e.g. in admin settings.

npretto commented on frontend/lint/tests/no-literal-metabase-strings.unit.spec.js:
```diff
@@ -0,0 +1,149 @@
+  {
+    // "Detect in literal strings",
+    code: `
+  import { getApplicationName } from 'metabase/selectors/whitelabel';
```
Why should we consider this as valid?
It allows this:
```tsx
import { t } from "ttag";
import { useSelector } from "metabase/lib/redux";
import { getApplicationName } from "metabase/selectors/whitelabel";

export function SomeComponent() {
  const appName = useSelector(getApplicationName);
  return (
    <div>
      <h1>{"Hello Metabase - hardcoded"}</h1>
      <h1>Metabase</h1>
      <h1>{t`Metabase`}</h1>
      <Something label="Metabase" />
      <Something label={t`Metabase`} />
      <h2>{appName}</h2>
    </div>
  );
}

const Something = ({ label }: { label: string }) => {
  return <h2>{label}</h2>;
};
```
where we us the applicationName once but can also hardcode it in other places

WiNloSt replied:
It's a compromise for the rule to not get too complicated. Surely we can try to track where an imported name is being used, but that would definitely require more time to implement.

This simplifies the implementation so that we trust if users import the selector they will use it correctly.

At least this should catch the majority of the problem where people just forget to import `getApplicationName` from the start. Actually that won't help when you're adding a new string in the same file that already has the selector imported though.

I didn't want to spend too much time on this so I think this is good enough.

npretto replied:
I'm still not understanding why we check if that import is present, isn't the name returned by that selector supposed to replace the hardcoded "Metabase" string?
ie, if before it was
```tsx
const Component = () => {

  return <h1>t`Welcome to Metabase`<h1>
}
 ```
 
 it should be fixed as
 ```tsx
import { useSelector } from "metabase/lib/redux";
import { getApplicationName } from "metabase/selectors/whitelabel";

const Component = () => {
   const applicationName = useSelector(getApplicationName); 

  return <h1>t`Welcome to ${applicationName}`<h1>
}
 ```
which removes "Metabase" from the source code.

Shouldn't we be able to remove the import check entirely?

WiNloSt replied:
Fixed in https://github.com/metabase/metabase/pull/38553/commits/9219375409a6a48697498ec1b8db7f852fd9cd07. This pretty much moves all the previous tests that are supposed to fail from `VALID_CASES` to `INVALID_CASES`

WiNloSt committed:
Merge branch 'master' into introduce-eslint-for-metabase-strings

WiNloSt committed:
Address review: Fix the rule

The rule was checking if we have imported the selector
`getApplicationName` then ignore all Metabase strings. This is different
than `no-unconditional-metabase-links-render` because in that rule, when
fixed the Documentation URLs are still in the file, but in this rule,
when fixed, there should be no Metabase strings left in the file.

WiNloSt committed:
Fix errors from the new lint rule

npretto reviewed (approved):
Tested and it works 👍 
Not an expert of eslint specific code but I know it is similar to the already approved PR for the links
